TLS Session ticket support

redbrain17 opened the pull request:
- fixed issues of review
- rebased with current master

don't take the two former pull request. I think I messed up these. Please review only this one

jay commented:
Closed #3058 #3059 in favor of this one. In the future if you need to change the commits please force-push to your branch rather than start a new pull request with a new branch. Session tickets have been proposed before but the idea usually petered off, I'm not sure why. Is there some advantage to using them over session id which is already supported?

bagder commented on lib/vtls/nss.c:
```diff
@@ -2029,6 +2029,11 @@ static CURLcode nss_setup_connect(struct connectdata *conn, int sockindex)
+   if (data->set.ssl.sessionticket) {
+           if (SSL_OptionSet(BACKEND->handle, SSL_ENABLE_SESSION_TICKETS, PR_TRUE) != SECSuccess) {
+               goto error;
+           }
```
wrong indent level

bagder commented on src/tool_getparam.c:
```diff
@@ -1659,6 +1660,10 @@ ParameterError getparameter(const char *flag, /* f or -long-flag */
 
+     case 'C': /* --tls-use-session-tickets */
+           config->tls_use_session_ticket = TRUE;
+		break;
```
wrong indent level

bagder commented on src/tool_help.c:
```diff
@@ -332,6 +332,8 @@ static const struct helptxt helptext[] = {
   {"    --proxy-tlsauthtype <type>",
    "TLS authentication type for HTTPS proxy"},
+  {"    --tls-use-session-tickets",
+   "Enable TLS session tickets (NSS)"},
```
note that this list is generated (see comment above the list), so the contents need to be in files in `docs/cmdline-opts/` - you need to create a new file there for a new option

redbrain17 replied:
ok I added a file tls-sessiontickets.d should I drop this two lines then?

bagder replied:
No, they still need to be there. but my comment about "(NSS)" applies here too...

redbrain17 replied:
Missed you comment according (NSS) So I dropped it. But now its not clear that RFC 5077 is only supported by NSS implementation of TLS or I'am wrong here?

bagder commented on src/tool_operate.c:
```diff
@@ -1482,6 +1482,10 @@ static CURLcode operate_do(struct GlobalConfig *global,
 
+	if (config->tls_use_session_ticket) {
+          my_setopt(curl, CURLOPT_TLS_USE_SESSION_TICKETS, 1L);
+        }
```
wrong indent level

redbrain17 replied:
do  `make checksrc` find such indent level errors?

bagder commented:
Could we consider simple using a ticket if `CURLOPT_SSL_SESSIONID_CACHE` is true instead of adding a new option? Will there really be a situation where a user wants one but not the other?

bagder commented:
... or perhaps convert it to a bitmask and add a bitflag for ticket?

bagder commented:
Also, pay extra attention to the [test 1119 and 1139 failures](https://travis-ci.org/curl/curl/jobs/434696188#L5827).

redbrain17 commented:
the difference between TLS session ids and TLS Session Tickets are explained here:

https://stackoverflow.com/questions/19939247/ssl-session-tickets-vs-session-ids
according to this link there exits indeed situations where tickets are desirable over session identifiers. (RFC 5077)
So I would suggest an extra option here.

jay commented:
Has anyone done a real-world comparison of the proportion of servers that support each method?

bagder commented:
> So I would suggest an extra option here.

What about instead making bit 1 of the argument to`CURLOPT_SSL_SESSIONID_CACHE` become a flag to toggle session tickets? The option only has a single bit defined currently so it would not break anything but will save us from adding another option. The session_id is then not entirely correctly named but is still a function that is very similar in function.

mkauf commented:
#2645 is another old version of this PR.

danielgustafsson commented:
Ping @redbrain17. Have you had time to work on addressing the comments made on this PR?

redbrain17 commented:
On 24/12/18, Daniel Gustafsson wrote:
> Ping @redbrain17. Have you had time to work on addressing the comments made on this PR?

still alive and interested to get this merged. I'll look in the next
days on it to improve it.  

Best Regards,

Oliver

danielgustafsson commented:
> On 26 Dec 2018, at 11:45, Oliver Graute <notifications@github.com> wrote:
> 
> On 24/12/18, Daniel Gustafsson wrote:
> > Ping @redbrain17. Have you had time to work on addressing the comments made on this PR?
> 
> still alive and interested to get this merged. I'll look in the next
> days on it to improve it. 

Great! Let us know if you need any help.

redbrain17 commented:
improved the patches. Ready for next review.

bagder commented on src/tool_getparam.c:
```diff
@@ -1670,6 +1671,11 @@ ParameterError getparameter(const char *flag, /* f or -long-flag */
 
+      case 'C':
+        /* --tls-use-session-tickets */
+        config->tls_use_session_ticket = TRUE;
```
this should be `= toggle` so that the `--no-` prefix works.

redbrain17 replied:
done

bagder commented on src/tool_getparam.c:
```diff
@@ -264,6 +264,7 @@ static const struct LongShort aliases[]= {
   {"E9", "proxy-tlsv1",              ARG_NONE},
   {"EA", "socks5-basic",             ARG_BOOL},
   {"EB", "socks5-gssapi",            ARG_BOOL},
+  {"EC", "tls-use-session-tickets",  FALSE},
```
Use `ARG_BOOL` instead of `FALSE`.

bagder commented on lib/vtls/nss.c:
```diff
@@ -2070,6 +2070,11 @@ static CURLcode nss_setup_connect(struct connectdata *conn, int sockindex)
 #endif
 
+   if (data->set.ssl.sessionticket) {
+           if (SSL_OptionSet(BACKEND->handle, SSL_ENABLE_SESSION_TICKETS, PR_TRUE) != SECSuccess) {
```
Is there a particular reason to have this feature default to off?

bagder commented on docs/cmdline-opts/tls-sessiontickets.d:
```diff
@@ -0,0 +1,7 @@
+Added: 7.6x.0
+---
+Set TLS Session Tickets in NSS for Session Resumption without Server-Side
+State (RFC 5077)
```
I think this explanation should expand a little on why a user would like to use this option even if they're not experts in TLS protocol details.

Perhaps drop "use" part from the flag name to shorten it somewhat?

redbrain17 replied:
done

bagder commented on include/curl/curl.h:
```diff
@@ -1894,6 +1894,9 @@ typedef enum {
   CINIT(HTTP09_ALLOWED, LONG, 285),
 
+  /* TLS session tickets*/
+  CINIT(TLS_USE_SESSION_TICKETS, LONG, 286),
```
I think perhaps this should rather be a new bit in the `CURLOPT_SSL_OPTIONS` option. What do you think?

It also needs to be documented. And added to `docs/libcurl/symbols-in-versions`.

bagder commented:
Abandoned. Closing.
